Added strings algorithms and it's test file #8

Aryamanz29 opened the pull request:
### Description

Added strings algorithms and a test file.

### Related Issue

Fixes #8 

### Type of Change:

- Feature

### How Has This Been Tested?

-Use python unittest framework for testing is file.

### Checklist:

- [x] I read the **CONTRIBUTING** document before making this pull request
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [x] I have updated documentation accordingly
- [x] My changes generate no new warnings 
- [x] Existing tests passed

Aryamanz29 commented:
> ### Description
> Added strings algorithms and a test file.
> 
> ### Related Issue
> Fixes #8
> 
> ### Type of Change:
> * Feature
> 
> ### How Has This Been Tested?
> -Use python unittest framework for testing is file.
> 
> ### Checklist:
> * [x]  I read the **CONTRIBUTING** document before making this pull request
> * [x]  My code follows the style guidelines of this project
> * [x]  I have performed a self-review of my own code or materials
> * [x]  I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
> * [x]  I have updated documentation accordingly
> * [x]  My changes generate no new warnings
> * [x] Existing tests passed

devkapilbansal commented:
@Aryamanz29 write steps for testing

Aryamanz29 commented:
> @Aryamanz29 write steps for testing
Steps for testing on terminal/shell
1.>>python
2.Import unittest from the standard library
3.Run python string_tests.py

Aryamanz29 commented:
> > @Aryamanz29 write steps for testing


> Steps for testing on terminal/shell
> 1.>>python
> 2.Import unittest from the standard library
> 3.Run python string_tests.py
>

Aryamanz29 commented:
> > > @Aryamanz29 write steps for testing


> > Steps for testing on terminal/shell
> > 1.>>python
> > 2.Import unittest from the standard library
> > 3.Run python string_tests.py

devkapilbansal commented on dsalgo/string_algo.py:
```diff
@@ -0,0 +1,158 @@
+
```
Why a blank line here??

devkapilbansal commented on dsalgo/string_algo.py:
```diff
@@ -0,0 +1,158 @@
+
+class Strings:
+
+# Implement the encode and decode methods.
```
Your comment doesn't seem relevant and it's not explains anything

devkapilbansal commented on dsalgo/string_algo.py:
```diff
@@ -0,0 +1,158 @@
+# Implement the encode and decode methods.
+     
+     def encode(self,strs):
+         """Encodes a list of strings to a single string.
```
Can you please use docstrings here??

devkapilbansal commented on tests/string_tests.py:
```diff
@@ -0,0 +1,25 @@
+
+import unittest
+from dsalgo.string_algo import Strings
+class TestEncodeDecode(unittest.TestCase):
```
Add a blank line after import command

devkapilbansal replied:
only one is required

devkapilbansal reviewed (changes requested):
Use docstrings everywhere

devkapilbansal commented on dsalgo/string_algo.py:
```diff
@@ -0,0 +1,156 @@
+            i = index + 1 + size
+        return strs
+
+    def fizzbuzz(self, n):
+        """
+        params : N int
+        returns: string
+
+        A function that returns an array containing
+        the numbers from 1 to N,
+        Where N (size of an array) is the parametered value.
+        N will never be less than 1.
+        There is no fancy algorithm to solve fizz buzz.
+        Iterate from 1 through n
+        Use the mod operator to determine if the current
+        iteration is divisible by:
+        3 and 5 -> 'FizzBuzz'
+        3 -> 'Fizz'
+        5 -> 'Buzz'
+        else -> string of current iteration
+        return the results
+        Complexity:
+        Time: O(n)
+        Space: O(n)
+        """
+        # Validate the input
+        if n < 1:
+            raise ValueError("n cannot be less than one")
+        if n is None:
+            raise TypeError("n cannot be None")
```
Remove this function. @codesankalp @ravi5175 I don't think that it is valuable to us in any sense. What are your views?

devkapilbansal commented on dsalgo/string_algo.py:
```diff
@@ -0,0 +1,156 @@
+class Strings:
+    def encode(self, strs):
+        """
+        Encodes a list of strings to a single string.
+        params: List[str]
+        returns: str
+        """
```
Not able to understand clearly what your function is doing by this. Please elaborate

devkapilbansal commented on dsalgo/string_algo.py:
```diff
@@ -0,0 +1,156 @@
+        return res
+
+    def decode(self, s):
+        """
+        Decodes a single string to a list of strings.
+        params: str
+        returns: List[str]
```
Similar to elaborate

devkapilbansal commented on dsalgo/string_algo.py:
```diff
@@ -0,0 +1,156 @@
+                result.append(i)
+        return result
+
+    def add_binary(self, a, b):
+        """
+        params:string
+        returns:string
+
+        Given two binary strings,
+        return their sum (also a binary string).
+        For example,
+        a = "11"
+        b = "1"
+        Return "100".
+        """
+        s = ""
+        c, i, j = 0, len(a) - 1, len(b) - 1
+        zero = ord("0")
+        while i >= 0 or j >= 0 or c == 1:
+            if i >= 0:
+                c += ord(a[i]) - zero
+                i -= 1
+            if j >= 0:
+                c += ord(b[j]) - zero
+                j -= 1
+            s = chr(c % 2 + zero) + s
+            c //= 2
+        return s
```
Isn't it a math algo?? Then why it is here

Aryamanz29 replied:
ohk I'll remove that one

devkapilbansal commented on tests/string_tests.py:
```diff
@@ -0,0 +1,29 @@
+
+    def test_roman(self):
+        self.assertAlmostEqual(9, self.st.roman_to_int("IX"))
+
+
+if __name__ == "__main__":
+    unittest.main()
```
Probably if we run this with unittest module, then we don't need this too

devkapilbansal commented:
Squash your commit and use proper commit guidelines

Commit:
[feature] Added string algorithms and tests. #8

Added below mentioned algos and their tests:

- Encode
- Decode
- repeat_substring
- longest_palindrome
- roman_to_int

Fixes #8
